### ISSUE TITLE
Cloud-init v. 23.3.1-0ubuntu1~20.04.1 running 'init' after update

### ISSUE THREAD

**Issue by t3easy** [canonical/cloud-init#4501]
# Bug report

After the update of an existing server to 23.3.1-0ubuntu1~20.04.1, cloud-init runs init and generates new ssh host keys and the user ubuntu. It seems like cloud-init 23.3.1 ignored the VMware OVF settings transported over VMware Tools.

## Steps to reproduce the problem

Update an existing ubuntu 20.04.1 server that was configuresd by OVF over VMware Tools to cloud init 23.3.1

## Environment details
- Cloud-init version: 23.3.1-0ubuntu1~20.04.1
- Operating System Distribution: Ubuntu
- Cloud provider, platform or installer type: VMware vSphere, vApp OVF over VMware Tools

**Comment by t3easy** [comment]
This also delays the reboot for 120 seconds, as it tries to get cloud init config from gateway ip.

**Comment by aciba90** [comment]
Thanks, @t3easy, for making cloud-init better. In order to successfully inspect what's going on, could you please provide the cloud-init [logs](https://canonical-cloud-init.readthedocs-hosted.com/en/latest/howto/bugs.html#collect-logs), making sure any sensible information is redacted?

**Comment by GeoffreyFrogeye** [comment]
We have a similar issue, but on Xen. I investigated and I think it's the same issue as @t3easy describes though.

Our VMs have `cloud-init` installed by default, but only explicitely enabled via a `ds=` kernel parameter when we need it. Affter upgrading `cloud-init` to `23.3.1-0ubuntu1~20.04.1`, the `cloud-init` systemd services now are automatically started on boot, and act as if it was the first boot, changing SSH host keys and doing some other stuff. But unlike when we set it automatically, it also delays boot by ~2 minutes.

`/lib/cloud-init/ds-identify` now returns `0` instead of `1` in the previous version, causing `cloud-init` services to start. `CloudStack` is now (wrongly) set as a data-source as seen in `cloud-init query --all`. I suspect the commit introducing changing the behaviour to be https://github.com/canonical/cloud-init/commit/7949bb3de352dc40c62c38a0534722b24b7eed22#diff-8e27dfc14b3ffceee1e29d3cf58dc25309c0ac070879febd47c578de04ea2dddR742-R744 , which now suggests that the VM might be running on `CloudStack` whenever it's running on VMware or Xen (hence why I think this is the same issue).

This is quite the behaviour change being introduced for a minor release, and thus in regular package upgrade. As a small cloud provider, this has quite the impact, as our customers keeping their machine automatically up to date and rebooting suddenly got their host SSH keys changed.

**Comment by TheRealFalcon** [comment]
@GeoffreyFrogeye and @t3easy , thank you for the information. I believe I understand the issue but still need some more information. If you could provide the result of `cloud-init collect-logs` here (with any sensitive data removed), that would be very helpful. I need to understand:
1. Had cloud-init previously run on the instance, or was it always disabled by ds-identify until this point
2. If not running CloudStack, why is cloud-init determining the VM to be CloudStack rather than DataSourceNone

Beyond this, I need to understand if this is behavior that would be problematic if it hadn't been introduced to an already running instance. The reason I ask is because I think that the change in 7949bb3 is still technically correct as CloudStack *may* be running on these virtualization platforms, so cloud-init should check this on first boot. The incorrect behavior was to apply this change to existing instances. Does that sound correct to you?

**Comment by t3easy** [comment]
@TheRealFalcon My VMs are deployed with the Ubuntu Server Cloud Image OVA and did run cloud-init on first boot already about two years ago.

We're not using CloudStack, just VMware vSphere VMs.

I can't provide logs right now.

**Comment by mikat-gh** [comment]
We are hitting this bug when installing new VMware VM from Ubuntu 22.04 Cloud image. **It completely breaks our automatic provisioning.**

We use the VMware ds for initial setup which includes:
```
package_upgrade: true
package_reboot_if_required: true
```
Now after `cloud-init` is updated during the initial boot, it runs again on second boot and resets our static network config to DHCP. Relevant parts of log:
```
<initial boot>
2023-10-16 06:21:58,975 - util.py[DEBUG]: Cloud-init v. 23.2.2-0ubuntu0~22.04.1 running 'init-local' at Mon, 16 Oct 2023 06:21:58 +0000. Up 13.86 seconds.
2023-10-16 06:21:58,975 - main.py[DEBUG]: No kernel command line url found.
2023-10-16 06:21:58,975 - main.py[DEBUG]: Closing stdin.
2023-10-16 06:21:58,979 - util.py[DEBUG]: Writing to /var/log/cloud-init.log - ab: [640] 0 bytes
2023-10-16 06:21:58,980 - util.py[DEBUG]: Changing the ownership of /var/log/cloud-init.log to 104:4
2023-10-16 06:21:58,980 - util.py[DEBUG]: Writing to /var/lib/cloud/data/python-version - wb: [644] 4 bytes
2023-10-16 06:21:58,981 - util.py[DEBUG]: Attempting to remove /var/lib/cloud/instance/boot-finished
2023-10-16 06:21:58,986 - handlers.py[DEBUG]: start: init-local/check-cache: attempting to read from cache [check]
2023-10-16 06:21:58,986 - util.py[DEBUG]: Reading from /var/lib/cloud/instance/obj.pkl (quiet=False)
2023-10-16 06:21:58,986 - stages.py[DEBUG]: no cache found
2023-10-16 06:21:58,986 - handlers.py[DEBUG]: finish: init-local/check-cache: SUCCESS: no cache found
2023-10-16 06:21:58,986 - util.py[DEBUG]: Attempting to remove /var/lib/cloud/instance
2023-10-16 06:21:59,015 - stages.py[DEBUG]: Using distro class <class 'cloudinit.distros.ubuntu.Distro'>
2023-10-16 06:21:59,016 - __init__.py[DEBUG]: Looking for data source in: ['VMware', 'None'], via packages ['', 'cloudinit.sources'] that matches dependencies ['FILESYSTEM']
...
2023-10-16 06:21:59,898 - stages.py[INFO]: Applying network configuration from ds bringup=False: {'version': 2, 'ethernets': {'ens192': {'addresses': ['XXX'], 'routes': [{'to': '0.0.0.0/0', 'via': 'XXX'}], 'nameservers': {'addresses': ['XXX', 'XXX']}}}}
<second boot after package_upgrade>
2023-10-16 06:25:38,210 - util.py[DEBUG]: Cloud-init v. 23.3.1-0ubuntu1~22.04.1 running 'init-local' at Mon, 16 Oct 2023 06:25:38 +0000. Up 12.51 seconds.
2023-10-16 06:25:38,211 - main.py[DEBUG]: No kernel command line url found.
2023-10-16 06:25:38,211 - main.py[DEBUG]: Closing stdin.
2023-10-16 06:25:38,218 - util.py[DEBUG]: Writing to /var/log/cloud-init.log - ab: [640] 0 bytes
2023-10-16 06:25:38,220 - util.py[DEBUG]: Changing the ownership of /var/log/cloud-init.log to 104:4
2023-10-16 06:25:38,221 - util.py[DEBUG]: Attempting to remove /var/lib/cloud/instance/boot-finished
2023-10-16 06:25:38,222 - handlers.py[DEBUG]: start: init-local/check-cache: attempting to read from cache [check]
2023-10-16 06:25:38,222 - util.py[DEBUG]: Reading from /var/lib/cloud/instance/obj.pkl (quiet=False)
2023-10-16 06:25:38,223 - util.py[DEBUG]: Read 9863 bytes from /var/lib/cloud/instance/obj.pkl
2023-10-16 06:25:38,262 - stages.py[DEBUG]: cache invalid in datasource: DataSourceVMware [seed=guestinfo]
2023-10-16 06:25:38,262 - handlers.py[DEBUG]: finish: init-local/check-cache: SUCCESS: cache invalid in datasource: DataSourceVMware [seed=guestinfo]
2023-10-16 06:25:38,262 - util.py[DEBUG]: Attempting to remove /var/lib/cloud/instance
2023-10-16 06:25:38,265 - stages.py[DEBUG]: Using distro class <class 'cloudinit.distros.ubuntu.Distro'>
2023-10-16 06:25:38,266 - __init__.py[DEBUG]: Looking for data source in: ['CloudStack', 'None'], via packages ['', 'cloudinit.sources'] that matches dependencies ['FILESYSTEM']
...
2023-10-16 06:25:38,320 - stages.py[INFO]: Applying network configuration from fallback bringup=False: {'ethernets': {'ens192': {'dhcp4': True, 'set-name': 'ens192', 'match': {'macaddress': 'XXX'}}}, 'version': 2}

**Comment by GeoffreyFrogeye** [comment]
@TheRealFalcon

1. It was always disabled by ds-identify until this point.
2. For you to tell us I guess, I made a conjecture in my first comment (tl;dr cloud-init thinks if a VM is running on Xen/VMware then it must be CloudStack ; from commenter on this issue this is not necessarily the case) but I am not familiar enough with cloud-init code to be 100% sure.

Here are the results of `cloud-init collect-logs` before and after upgrading `cloud-init` to 23.3 and rebooting on a fresh instance of Ubuntu 22.04 (the OS template being slightly outdated, hence why it shipped with 23.2): [before.tar.gz](https://github.com/canonical/cloud-init/files/12916393/before.tar.gz) [after.tar.gz](https://github.com/canonical/cloud-init/files/12916400/after.tar.gz)

> The incorrect behavior was to apply this change to existing instances.

No, I believe that a fresh instance shipped with 23.3 would have the same problematic behaviour, notably the boot delays. That those changes were applied to existing instances did make the problem worse though.

**Comment by TheRealFalcon** [comment]
As a short term workaround to the issues mentioned here, you can either disable cloud-init entirely after first boot, or force the VMWare datasource every boot.

1. To disable cloud-init after first boot, the easiest method would be to `touch /etc/cloud/cloud-init.disabled`, though any of the methods listed [in the docs](https://cloudinit.readthedocs.io/en/latest/howto/disable_cloud_init.html) should work.
2. To force usage of the VMWare datasource on Ubuntu, you can `echo 'datasource_list: [ VMware ]' > /etc/cloud/cloud.cfg.d/90_dpkg.cfg `

@GeoffreyFrogeye , thanks for the logs. I understand the problem and we'll work on getting a fix up soon.

@mikat-gh , the symptoms of your issue are different, and I still don't have enough information to understand the root cause. The commit mentioned in an earlier comment (7949bb3d) should only cause problems if no datasource is detected. It appears your system was initially detected as VMWare but is no longer being detected as so. There are legitimate reasons that could have happened (i.e., the VMWare tools used on first boot no longer being available), but nothing in 7949bb3d would have caused the VM to no longer be detected as VMWare, so I'm not sure that the fix being proposed here will solve your problem. If your VM is happy to live the rest of its days as DataSourceNone after first boot, then the fix mentioned here should be fine. If not, please open a separate issue containing the full logs from `cloud-init collect-logs` (with any sensitive info redacted), and we'll try to get to the bottom of it.

**Comment by mikat-gh** [comment]
@TheRealFalcon Reason for not detecting VMware on second run is probably caused by us using [redact](https://cloudinit.readthedocs.io/en/latest/reference/datasources/vmware.html#redacting-sensitive-information-guestinfo-keys-transport-only), so VMware config is no longer available on the second run. We actually disable cloud-init on later stage of the setup with Ansible.

Thanks for looking into it!
